### PR TITLE
feat: #3540 Ability to preserve marks on lists

### LINKED ISSUE #3540: Ability to preserve marks from previous paragraph when toggling a listItem
### What problem are you facing?

We are in need of this functionality similar to `msWord`.

### What’s the solution you would like to see?

We have started working on it, similar to hardbreak extension, we plan to use a `keepMarks` optional property on the `splitList` command when can use to preserve marks from the previous line. Hoping to raise a PR for this shortly.

### What alternatives did you consider?
_No response_
### Anything to add? (optional)

_No response_

### Are you sponsoring us?

- [ ] Yes, I’m a sponsor. 💖

### PR DESCRIPTION
**Pull request by gethari**
Fixes #3540 

Attaching a sample clip on the demo.

- When we have a paragraph with marks, and when the user moves to next line and toggle any list item, the marks can be preserved based on the `keepMarks` flag similar to `hardBreak` extension

https://user-images.githubusercontent.com/20477956/207530906-6abbb4a1-67a5-4b1f-9944-310ad4055c8a.mp4

### CODE REVIEW

**Comment by gethari** [comment]
@bdbch can you please review this when you get a chance, also do let me know if it would be ideal to do the same `keepMarks` for `TaskList` as well 🚀

File: packages/core/src/inputRules/wrappingInputRule.ts [hunk 1]
```diff
@@ -45,6 +51,24 @@ export function wrappingInputRule(config: {
...
+
+        if (marks) {
+          const filteredMarks = marks.filter(mark => splittableMarks.includes(mark.type.name))
+
+          tr.ensureMarks(filteredMarks)
+        }
+      }
+      if (config.keepAttributes) {
+        /** If the nodeType is `bulletList` or `orderedList` set the `nodeType` as `listItem` */
+        const nodeType = config.type.name === 'bulletList' || config.type.name === 'orderedList' ? 'listItem' : 'taskList'
```
Comment (gethari) [inline packages/core/src/inputRules/wrappingInputRule.ts:67]: @bdbch I'm just worried about this piece of code, please let me know if there's any better way of doing this. The idea here is 
- When there's a paragraph with say colored text
- When we switch over to 2nd line, we apply the color to the whole `<li>` only then it is color is also applied for the `list-style` indicator.

**Comment by bdbch** [comment]
Thanks @gethari I'll take a look as soon as I can! 💪

**Comment by gethari** [comment]
@bdbch did you get a chance to check this out ? Please share your thoughts / feedback

**Comment by gethari** [comment]
@bdbch , I hope this message finds you well. I wanted to inform you that I have recently updated the branch to reflect the latest fork and resolved any conflicts that arose. I would be grateful if you could take a moment to review these changes at your earliest convenience. Thank you!

**Comment by bdbch** [comment]
Sure thing. I'll check your PR out as soon as I find some time.

**Review by bdbch — APPROVED** [review]

File: demos/src/Examples/Default/React/index.jsx [hunk 2]
```diff
@@ -165,14 +168,31 @@ const MenuBar = ({ editor }) => {
...
 export default () => {
   const editor = useEditor({
     extensions: [
-      StarterKit,
+      Color.configure({ types: [TextStyle.name, ListItem.name] }),
+      TextStyle.configure({ types: [ListItem.name] }),
+      StarterKit.configure({
+        bulletList: {
+          keepMarks: true,
+          keepAttributes: false, // TODO : Making this as `false` becase marks are not preserved when I try to preserve attrs, awaiting a bit of help
```
[the remark this answers is not in the document]
Comment (gethari) [reply demos/src/Examples/Default/React/index.jsx:189]: Let's say I have my first line with `Purple` color  and `Italic` mark applied. When we set `keepAttributes` to `true` I am trying to update it with whatever attributes that are there from the previous line. But while doing so, if we have any mark it is not being preserved. But even without keeping the `keepAttributes` to `true` the color from the `<p>` is being preserved on the bullet list.


[In this line](https://github.com/gethari/tiptap/blob/50b95b938ff0c6d496414f7b08713bc4fe95e5bc/packages/extension-bullet-list/src/bullet-list.ts#L58]

```ts
 return chain().toggleList(this.name, this.options.itemTypeName, this.options.keepMarks).updateAttributes(ListItem.name, this.editor.getAttributes(TextStyle.name)).run()
```

[the remark this answers is not in the document]
Comment (gethari) [reply demos/src/Examples/Default/React/index.jsx:189]: My doubt is do I need to chain the `updateAttributes` again ?

[the remark this answers is not in the document]
Comment (bdbch) [reply demos/src/Examples/Default/React/index.jsx:189]: I think so - but shouldn't be a problem. Thanks for coming up with more information! I think this should also close #3768

File: packages/extension-ordered-list/src/ordered-list.ts [hunk 3]
```diff
@@ -78,13 +88,23 @@ export const OrderedList = Node.create<OrderedListOptions>({
...
+      find: inputRegex,
+      type: this.type,
+    })
+
+    if (this.options.keepMarks || this.options.keepAttributes) {
+      inputRule = wrappingInputRule({
         find: inputRegex,
         type: this.type,
-        getAttributes: match => ({ start: +match[1] }),
-        joinPredicate: (match, node) => node.childCount + node.attrs.start === +match[1],
```
Comment (javan) [inline packages/extension-ordered-list/src/ordered-list.ts:86]: I believe these removed lines broke support for creating ordered lists with a `start` offset and have opened an issue with more details: https://github.com/ueberdosis/tiptap/issues/3831

Comment (gethari) [reply packages/extension-ordered-list/src/ordered-list.ts:86]: @javan great find, I have raised #3833 to resolve this

**Pull request merged**